### ISSUE TITLE
tf.gradient returns None 

### ISSUE THREAD

**Issue by DagonArises** [tensorflow/tensorflow#55262]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Big Sur 11.3.1
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): v2.6.1-9-gc2363d6d025 2.6.2
- Python version: Python 3.6.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
In training an LSTM model I would like to compute the gradients of the cell state at step t on the initial cell state: dc_t/dc_0. However calling ```tf.gradients(ct, c0)``` returns None type output.

**Describe the expected behavior**
My c_t c_0 are both of shape (100, 200). I am expecting the gradient to be a real-valued Tensor of the same shape. 

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
### 1. Define model
batch_size = 100
input_length_m= 1403
output_dim= 100

xtr_pad = tf.random.uniform((batch_size, input_length_m), maxval = 500, dtype=tf.int32)
ytr = tf.random.normal((batch_size, input_length_m, 200))
inp= Input(batch_shape= (batch_size, input_length_m), name= 'input') 
emb_out= Embedding(500, output_dim, input_length= input_length_m, trainable= False, name= 'embedding')(inp)

class LSTMCellwithStates(LSTMCell):
    def call(self, inputs, states, training=None):
        real_inputs = inputs[:,:self.units] # decouple [h, c]
        outputs, [h,c] = super().call(real_inputs, states, training=training)
        return tf.concat([h, c], axis=1), [h,c]
    
rnn = RNN(LSTMCellwithStates(200), return_sequences= True, return_state= False, name= 'LSTM') 
h0 = tf.Variable(tf.random.uniform((batch_size, 200)))
c0 = tf.Variable(tf.random.uniform((batch_size, 200)))
rnn_allstates= rnn(emb_out, initial_state=[h0, c0])  
model_lstm_mod = Model(inputs=inp, outputs= rnn_allstates, name= 'model_LSTMCell')
model_lstm_mod.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['acc'])

### 2. Compute gradients:
ds = tf.data.Dataset.from_tensor_slices((xtr_pad, ytr)).batch(100)

@tf.function
def compute_dct_dc0(ct, c0):
    return tf.gradients(ct, c0)

n_b = int(xtr_pad.shape[0]/ 100)
n_steps = 20   # look up only the first and last 20 steps

dctdc0_all= tf.zeros([n_b, n_steps])
for b, (x_batch_train, y_batch_train) in enumerate(ds):  
    grad_batch= []   
    cell_states= model_lstm_mod(x_batch_train)[:, :, 200:]
    for t in range(n_steps):  
        ct= cell_states[:, t, :]
        print(ct)
        # steps 0,...,19
        dctdc0_b_t = compute_dct_dc0(ct, c0)  # (batch_size, n_units)
        print(dctdc0_b_t)
        grad_t = tf.reduce_mean(abs(dctdc0_b_t[0]), [0,1]) # Scalar dctdc0 at the current batch and step
        print('step', t+1, 'of batch' ,b+1, 'done')
        grad_batch.append(grad_t)
    
    dctdc0_all= tf.concat([dctdc0_all, [grad_batch]], axis = 0)   
```
The error occurs specifically here:
```
### 3. 
@tf.function
def compute_dct_dc0(ct, c0):
    print(ct)
    print(c0)
    return tf.gradients(ct, c0)
compute_dct_dc0(ct, c0)

>> Tensor("ct:0", shape=(100, 200), dtype=float32)
<tf.Variable 'Variable:0' shape=(100, 200) dtype=float32>
[None]
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

Surprisingly when the function is defined in a slightly different manner the gradients are computed without issue:
```
@tf.function
# Compute gradients
def compute_dct_dc0(t, x, c0):
    return tf.gradients(model_lstm_mod(x)[:,t,:], c0)

n_b = int(xtr_pad.shape[0]/ 100) 
n_steps = 20   # look up only the first and last 20 steps

dctdc0_all= tf.zeros([n_b, n_steps])
for b, (x_batch_train, y_batch_train) in enumerate(ds):  # batches 0,1
    grad_batch= []   # a list of 1403 scalar gradients on the current batch
    for t in range(n_steps):  
        # steps 0,...,19
        dctdc0_b_t = compute_dct_dc0(t, x_batch_train, c0)  # (batch_size, n_units)
        grad_t = tf.reduce_mean(abs(dctdc0_b_t[0]), [0,1]) # Scalar dctdc0 at the current batch and step
        print('step', t+1, 'of batch' ,b+1, 'done')
        grad_batch.append(grad_t)
    
    dctdc0_all= tf.concat([dctdc0_all, [grad_batch]], axis = 0)   

## Outputs:
>> step 1 of batch 1 done
step 2 of batch 1 done
step 3 of batch 1 done
```
However this way is a lot more expensive because it will be fitted the entire batch for each step t. 
```cell_states= model_lstm_mod(x_batch_train)[:, :, 200:]``` fits the batch only once and extracts c_t for each t to compute gradients.

**Comment by DagonArises** [comment]
Another related question is that, is ```tf.gradients(ht h0)``` watching h0 over the computations? 
Gradient Tape is capable of watching the input but only the input. Using Gradient Tape while specifying ```tape.watch(h0)``` also returns None in gradients

**Comment by tilakrayal** [comment]
@DagonArises ,
I was facing different error while executing the give code.Please find the gist [here](https://colab.research.google.com/gist/tilakrayal/0cbffeb86c0f88560a633e847430044c/untitled250.ipynb) and share required dependencies or share the colab gist.Thanks!

**Comment by DagonArises** [comment]
@tilakrayal Sorry I forgot to specify the imports. Here is the [notebook](https://colab.research.google.com/drive/1CQyLVcz3VX6eEi1d5UqwU68HevkmHKg6#scrollTo=z8tmyVsRCwhb)

**Comment by tilakrayal** [comment]
@DagonArises ,
I do not have access to the link you have provided. Could you please provide the required permissions to view the files.
